apps/docs/pages/guides/auth/auth-postgres-hooks.mdx

```diff
@@ -0,0 +1,669 @@
+
+You can do this by creating a table such as:
+
+``sql
+create table public.password_failed_verification_attempts (
+	user_id uuid not null,
+  last_failed_at timestamp not null default now(),
+  primary key (user_id, factor_id)
+);
+```
+
+Then you should create a hook which will read and enter information into this table. Example:
+
+```sql
+create function public.hook_password_verification_attempt(event jsonb)
+	returns jsonb
+  language plpgsql
+  as $$
+	declare
+		last_failed_at timestamp;
+	begin
+		if event->'valid' is true
+		then
+			-- code is valid, accept it
+			return jsonb_build_object('decision', 'continue');
+		end if;
+
+		select last_failed_at into last_failed_at
+			from public.password_failed_verification_attempts
+			where
+				user_id = event->'user_id'
+
+		if last_failed_at is not null and now() - last_failed_at < interval '10 seconds'
+		then
+			-- last attempt was done too quickly
+			return jsonb_build_object(
+				'error', jsonb_build_object(
+					'http_code', 429,
+					'message',   'Please wait a moment before trying again.'));
+		end if;
+
+		-- record this failed attempt
+		insert into public.password_failed_verification_attempts
+			(
+				user_id,
+				last_failed_at
+			)
+			values
+			(
+				event->'user_id',
+				now()
+			)
+			on conflict
+        do update
+          set last_failed_at = now();
+
+		-- finally let Supabase Auth do the default behavior for a failed attempt
+		return jsonb_build_object('decision', 'continue');
+	end;
+	$$;
+```
+
+Supabase Auth will invoke this function with the `supabase_auth_admin` role which does not have permissions to modify the `public` schema. So you should grant it appropriate permissions:
+
+```sql
+grant execute
+	on function public.hook_password_verification_attempt
+	to supabase_auth_admin;
+
+grant all
+	on table public.password_failed_verification_attempts
+	to supabase_auth_admin;
+```
+
+Finally to make sure this function and table are not accessible from Supabase Serverless APIs, revoke those permissions from the `authenticated` and `anon` roles:
+
+```sql
+revoke execute
+	on function public.hook_password_verification_attempt
+	from authenticated, anon;
+
+revoke all
+	on table public.password_failed_verification_attempts
+	from authenticated, anon;
+```
+</Accordion.Item>
+</div>
+</Accordion>
+
+<Accordion
+ type="default"
+  openBehaviour="multiple"
+  chevronAlign="right"
+  justified
+  size="medium"
+  className="text-foreground-light">
+<div className="border-b pb-3">
+### Example: Add Custom Claims to Token
+<Accordion.Item
+        header={<span className="text-foreground font-bold">Details </span>}
+        id={`add-custom-claims-to-token`}
+      >
+
+****Scenario****
+
+You wish to assign roles to users. In particular, you wish to assign users with a `@supabase.com` email account a role of admin so that they can perform restricted actions
```

Comment:
Maybe be more direct in the scenario. You can word it something like:

Often you find yourself needing to assign different permissions to different users. One way of doing this is by adding a custom `admin` boolean property in the `user_metadata` section of the JWT, which can then be picked up by RLS policies or your backend APIs.

...

Then, instead of using the email-domain approach, which I don't think is very useful in general, maybe you can fetch from a table `public.admins` which has a `user_id` column referencing if the user is an admin or not.